File: Compass/Agents/Agent.cs
namespace Compass.Agents;

public abstract class Agent
{
    public abstract string Name { get; }
    public abstract Task<AgentOutput> Execute(string prompt, string model, string workingDirectory);
    public abstract Task EnsureLogin();
    
    public static Agent Create(Types agentType)
    {
        return agentType switch
        {
            Types.GithubCopilot => new GithubCopilot(),
            Types.Codex => new Codex(),
            Types.OpenCode => new OpenCode(),
            _ => throw new ArgumentException("Unsupported agent type: " + agentType),
        };
    }

    public enum Types
    {
        GithubCopilot,
        Codex,
        OpenCode,
    }
}

File: src/agents/agent.ts
import type { AgentOutput } from "../models";

export interface AgentOptions {
    allowFullAccess: boolean;
}

export const defaultAgentOptions: AgentOptions = {
    allowFullAccess: false
};

export abstract class Agent {
    abstract execute(prompt: string, model: string, workingDirectory: string): Promise<AgentOutput>;
    abstract init(): Promise<void>;
    abstract requiredBinaries(): string[];

    name: string = "Agent";
    options: AgentOptions;

    constructor(name: string, options: AgentOptions) {
        this.name = name;
        this.options = options;
    }
}